sql/src/main/java/org/apache/druid/sql/SqlLifecycle.java

```diff
@@ -280,22 +277,20 @@ private void checkAccess(Access access)
    */
   public PrepareResult prepare() throws RelConversionException
   {
-    synchronized (lock) {
-      if (state != State.AUTHORIZED) {
-        throw new ISE("Cannot prepare because current state[%s] is not [%s].", state, State.AUTHORIZED);
-      }
-      Preconditions.checkNotNull(plannerContext, "Cannot prepare, plannerContext is null");
-      try (DruidPlanner planner = plannerFactory.createPlannerWithContext(plannerContext)) {
-        this.prepareResult = planner.prepare(sql);
-        return prepareResult;
-      }
-      // we can't collapse catch clauses since SqlPlanningException has type-sensitive constructors.
-      catch (SqlParseException e) {
-        throw new SqlPlanningException(e);
-      }
-      catch (ValidationException e) {
-        throw new SqlPlanningException(e);
-      }
+    if (state != State.AUTHORIZED) {
```

Comment:
since `prepare` doesn't transition state, it might be nice to check for cancelled here so we don't do silly stuff.

I guess alternatively we could have added `State.CANCELLED`, since then all of the state transitions would naturally fail because the expected previous state would not match expectations and `transition` wouldn't have to be a boolean that callers check. This was seems ok too though, so up to you if you want to make `CANCELLED` a real state or not.

Reply:
Oh, i guess if `CANCELLED` was a real state you would need the locking on state which might complicate things a bit in a different way than cancel being a standalone thing.

Reply:
I added a new final state, `CANCELLED`.